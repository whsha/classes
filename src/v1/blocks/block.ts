/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

/** An enum containing the daily blocks */
export enum Block {
    A = "a",
    B = "b",
    C = "c",
    D = "d",
    E = "e",
    F = "f"
}