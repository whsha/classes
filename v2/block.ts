/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

/** An enum containing the daily blocks */
export enum Block {
    A = "a",
    B = "b",
    C = "c",
    D = "d",
    E = "e",
    F = "f"
}

/** The block in which lunch meets (on normal days) */
export const LunchBlock = Block.D;

/** Colors of the blocks */
export enum BlockColor {
    Red = "Red",
    Orange = "Orange",
    Yellow = "Yellow",
    Green = "Green",
    Blue = "Blue",
    Purple = "Purple",
    Tan = "Tan",
    None = "No Color"
}
