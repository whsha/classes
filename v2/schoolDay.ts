/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

/** A numeric value denoting the school day in the 7 day cycle */
export enum SchoolDay {
    One = 1,
    Two = 2,
    Three = 3,
    Four = 4,
    Five = 5,
    Six = 6,
    Seven = 7
}

/** An array of all of the school days */
export const SCHOOL_DAYS = Object.values(SchoolDay).filter(key => isNaN(Number(SchoolDay[key as unknown as keyof typeof SchoolDay]))) as unknown as SchoolDay[];