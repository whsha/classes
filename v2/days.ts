/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

import { Block, BlockColor, LunchBlock } from "./block";
import { SchoolDay } from "./schoolDay";

/** A map of the blocks to their schooldays to their colors */
type BlockDayColorMap = {
    readonly [B in Block]: {
        [D in SchoolDay]: BlockColor;
    }
};

/** A map of block colors to their school days and the blocks they meet then */
type ColorSchoolDayBlockMap = {
    readonly [B in BlockColor]: {
        [S in SchoolDay]?: Block;
    }
};

/** The map of blocks to their schooldays and block colors */
// tslint:disable: object-literal-sort-keys
const MapOfBlocksToColor: BlockDayColorMap = {
    [Block.A]: {
        [SchoolDay.One]: BlockColor.None,
        [SchoolDay.Two]: BlockColor.None,
        [SchoolDay.Three]: BlockColor.None,
        [SchoolDay.Four]: BlockColor.None,
        [SchoolDay.Five]: BlockColor.None,
        [SchoolDay.Six]: BlockColor.None,
        [SchoolDay.Seven]: BlockColor.None
    },
    [Block.B]: {
        [SchoolDay.One]: BlockColor.Orange,
        [SchoolDay.Two]: BlockColor.Yellow,
        [SchoolDay.Three]: BlockColor.Green,
        [SchoolDay.Four]: BlockColor.Tan,
        [SchoolDay.Five]: BlockColor.Red,
        [SchoolDay.Six]: BlockColor.Purple,
        [SchoolDay.Seven]: BlockColor.Blue
    },
    [Block.C]: {
        [SchoolDay.One]: BlockColor.Yellow,
        [SchoolDay.Two]: BlockColor.Orange,
        [SchoolDay.Three]: BlockColor.Orange,
        [SchoolDay.Four]: BlockColor.Yellow,
        [SchoolDay.Five]: BlockColor.Orange,
        [SchoolDay.Six]: BlockColor.Orange,
        [SchoolDay.Seven]: BlockColor.Yellow
    },
    [Block.D]: {
        [SchoolDay.One]: BlockColor.Green,
        [SchoolDay.Two]: BlockColor.Tan,
        [SchoolDay.Three]: BlockColor.Tan,
        [SchoolDay.Four]: BlockColor.Green,
        [SchoolDay.Five]: BlockColor.Yellow,
        [SchoolDay.Six]: BlockColor.Green,
        [SchoolDay.Seven]: BlockColor.Green
    },
    [Block.E]: {
        [SchoolDay.One]: BlockColor.Red,
        [SchoolDay.Two]: BlockColor.Red,
        [SchoolDay.Three]: BlockColor.Purple,
        [SchoolDay.Four]: BlockColor.Red,
        [SchoolDay.Five]: BlockColor.Purple,
        [SchoolDay.Six]: BlockColor.Tan,
        [SchoolDay.Seven]: BlockColor.Tan
    },
    [Block.F]: {
        [SchoolDay.One]: BlockColor.Blue,
        [SchoolDay.Two]: BlockColor.Purple,
        [SchoolDay.Three]: BlockColor.Blue,
        [SchoolDay.Four]: BlockColor.Blue,
        [SchoolDay.Five]: BlockColor.Blue,
        [SchoolDay.Six]: BlockColor.Red,
        [SchoolDay.Seven]: BlockColor.Purple
    }
};
// tslint:enable: object-literal-sort-keys

/** A map of block colors to the school days they meet on aswell as the blocks for said day */
const MapOfColorsToSchoolDaysToBlocks: ColorSchoolDayBlockMap = {
    [BlockColor.None]: {
        [SchoolDay.One]: Block.A,
        [SchoolDay.Two]: Block.A,
        [SchoolDay.Three]: Block.A,
        [SchoolDay.Four]: Block.A,
        [SchoolDay.Five]: Block.A,
        [SchoolDay.Six]: Block.A,
        [SchoolDay.Seven]: Block.A
    },
    [BlockColor.Red]: {
        [SchoolDay.One]: Block.E,
        [SchoolDay.Two]: Block.E,
        [SchoolDay.Four]: Block.E,
        [SchoolDay.Five]: Block.B,
        [SchoolDay.Six]: Block.F,
    },
    [BlockColor.Orange]: {
        [SchoolDay.One]: Block.B,
        [SchoolDay.Two]: Block.C,
        [SchoolDay.Three]: Block.C,
        [SchoolDay.Five]: Block.C,
        [SchoolDay.Six]: Block.C,
    },
    [BlockColor.Yellow]: {
        [SchoolDay.One]: Block.C,
        [SchoolDay.Two]: Block.B,
        [SchoolDay.Four]: Block.C,
        [SchoolDay.Five]: Block.D,
        [SchoolDay.Seven]: Block.C
    },
    [BlockColor.Green]: {
        [SchoolDay.One]: Block.D,
        [SchoolDay.Three]: Block.B,
        [SchoolDay.Four]: Block.D,
        [SchoolDay.Six]: Block.D,
        [SchoolDay.Seven]: Block.D
    },
    [BlockColor.Blue]: {
        [SchoolDay.One]: Block.F,
        [SchoolDay.Three]: Block.F,
        [SchoolDay.Four]: Block.F,
        [SchoolDay.Five]: Block.F,
        [SchoolDay.Seven]: Block.B
    },
    [BlockColor.Tan]: {
        [SchoolDay.Two]: Block.D,
        [SchoolDay.Three]: Block.D,
        [SchoolDay.Four]: Block.B,
        [SchoolDay.Six]: Block.E,
        [SchoolDay.Seven]: Block.E
    },
    [BlockColor.Purple]: {
        [SchoolDay.Two]: Block.F,
        [SchoolDay.Three]: Block.E,
        [SchoolDay.Five]: Block.E,
        [SchoolDay.Six]: Block.B,
        [SchoolDay.Seven]: Block.F
    }
};

/** The block colors that meet on a day mapped by their block */
export type BlockColorsForDay = {
    [B in Block]: BlockDayColorMap[B][SchoolDay]
};

/** Function to get the block colors that meet on a specific school day */
export function getBlockColorsForDay(day: SchoolDay): BlockColorsForDay {
    return {
        [Block.A]: MapOfBlocksToColor[Block.A][day],
        [Block.B]: MapOfBlocksToColor[Block.B][day],
        [Block.C]: MapOfBlocksToColor[Block.C][day],
        [Block.D]: MapOfBlocksToColor[Block.D][day],
        [Block.E]: MapOfBlocksToColor[Block.E][day],
        [Block.F]: MapOfBlocksToColor[Block.F][day]
    };
}

/** Function to get the block color that has a lab on the given day */
export function getLabBlockColorForDay(day: SchoolDay) {
    return MapOfBlocksToColor[Block.B][day];
}

/** Function to get the block color that has a lunch on the given day */
export function getLunchBlockColorForDay(day: SchoolDay) {
    return MapOfBlocksToColor[LunchBlock][day];
}

/** Function to get the school days that have a specific color */
export function getSchoolDaysThatHaveColor(color: BlockColor): SchoolDay[] {
    return Object.keys(MapOfColorsToSchoolDaysToBlocks[color])
        .map(x => parseInt(x, 10)) as unknown[] as SchoolDay[];
}

/** Checks if a color meets on a given day */
export function colorMeetsOnDay(color: BlockColor, day: SchoolDay) {
    return MapOfColorsToSchoolDaysToBlocks[color][day] !== undefined;
}

/** Function to get the block that the class meets having the given block color on the given day */
export function getBlockForColorOnDay(color: BlockColor, day: SchoolDay) {
    return MapOfColorsToSchoolDaysToBlocks[color][day] as unknown as Block;
}