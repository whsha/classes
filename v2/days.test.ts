/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

import { BlockColor } from "./block";
import { getBlockColorsForDay, getLabBlockColorForDay, getLunchBlockColorForDay } from "./days";
import { SchoolDay } from "./schoolDay";

describe("Tests utility functions for days", () => {
    it("Correctly calculates getBlockColorsForDay", () => {
        expect(getBlockColorsForDay(SchoolDay.One)).toMatchSnapshot();
        expect(getBlockColorsForDay(SchoolDay.Two)).toMatchSnapshot();
        expect(getBlockColorsForDay(SchoolDay.Three)).toMatchSnapshot();
        expect(getBlockColorsForDay(SchoolDay.Four)).toMatchSnapshot();
        expect(getBlockColorsForDay(SchoolDay.Five)).toMatchSnapshot();
        expect(getBlockColorsForDay(SchoolDay.Six)).toMatchSnapshot();
        expect(getBlockColorsForDay(SchoolDay.Seven)).toMatchSnapshot();
    });

    it("Correctly calculates getLabBlockColorForDay", () => {
        expect(getLabBlockColorForDay(SchoolDay.One)).toStrictEqual(BlockColor.Orange);
        expect(getLabBlockColorForDay(SchoolDay.Five)).toStrictEqual(BlockColor.Red);
        expect(getLabBlockColorForDay(SchoolDay.Seven)).toStrictEqual(BlockColor.Blue);
    });

    it("Correctly calculates getLunchBlockColorForDay", () => {
        expect(getLunchBlockColorForDay(SchoolDay.One)).toStrictEqual(BlockColor.Green);
        expect(getLunchBlockColorForDay(SchoolDay.Five)).toStrictEqual(BlockColor.Yellow);
        expect(getLunchBlockColorForDay(SchoolDay.Seven)).toStrictEqual(BlockColor.Green);
    });
});