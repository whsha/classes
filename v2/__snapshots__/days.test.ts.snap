// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests utility functions for days Correctly calculates getBlockColorsForDay 1`] = `
Object {
  "a": "No Color",
  "b": "Orange",
  "c": "Yellow",
  "d": "Green",
  "e": "Red",
  "f": "Blue",
}
`;

exports[`Tests utility functions for days Correctly calculates getBlockColorsForDay 2`] = `
Object {
  "a": "No Color",
  "b": "Yellow",
  "c": "Orange",
  "d": "Tan",
  "e": "Red",
  "f": "Purple",
}
`;

exports[`Tests utility functions for days Correctly calculates getBlockColorsForDay 3`] = `
Object {
  "a": "No Color",
  "b": "Green",
  "c": "Orange",
  "d": "Tan",
  "e": "Purple",
  "f": "Blue",
}
`;

exports[`Tests utility functions for days Correctly calculates getBlockColorsForDay 4`] = `
Object {
  "a": "No Color",
  "b": "Tan",
  "c": "Yellow",
  "d": "Green",
  "e": "Red",
  "f": "Blue",
}
`;

exports[`Tests utility functions for days Correctly calculates getBlockColorsForDay 5`] = `
Object {
  "a": "No Color",
  "b": "Red",
  "c": "Orange",
  "d": "Yellow",
  "e": "Purple",
  "f": "Blue",
}
`;

exports[`Tests utility functions for days Correctly calculates getBlockColorsForDay 6`] = `
Object {
  "a": "No Color",
  "b": "Purple",
  "c": "Orange",
  "d": "Green",
  "e": "Tan",
  "f": "Red",
}
`;

exports[`Tests utility functions for days Correctly calculates getBlockColorsForDay 7`] = `
Object {
  "a": "No Color",
  "b": "Blue",
  "c": "Yellow",
  "d": "Green",
  "e": "Tan",
  "f": "Purple",
}
`;

exports[`Tests utility functions for days Correctly calculates getSchoolDaysThatHaveColor 1`] = `
Array [
  1,
  2,
  4,
  5,
  6,
]
`;

exports[`Tests utility functions for days Correctly calculates getSchoolDaysThatHaveColor 2`] = `
Array [
  1,
  2,
  3,
  5,
  6,
]
`;

exports[`Tests utility functions for days Correctly calculates getSchoolDaysThatHaveColor 3`] = `
Array [
  1,
  2,
  4,
  5,
  7,
]
`;

exports[`Tests utility functions for days Correctly calculates getSchoolDaysThatHaveColor 4`] = `
Array [
  1,
  3,
  4,
  6,
  7,
]
`;

exports[`Tests utility functions for days Correctly calculates getSchoolDaysThatHaveColor 5`] = `
Array [
  1,
  3,
  4,
  5,
  7,
]
`;

exports[`Tests utility functions for days Correctly calculates getSchoolDaysThatHaveColor 6`] = `
Array [
  2,
  3,
  5,
  6,
  7,
]
`;

exports[`Tests utility functions for days Correctly calculates getSchoolDaysThatHaveColor 7`] = `
Array [
  2,
  3,
  4,
  6,
  7,
]
`;
